### PR TITLE
[Feature] Support AWQ 

### PR DESCRIPTION
**Pull request by pppppM**

### CODE REVIEW

File: README_zh-CN.md [hunk 1]
```diff
@@ -151,6 +151,30 @@ deepspeed --module --num_gpus 2 lmdeploy.pytorch.chat \
...
+
+```
+#   HF_model        DECODER_LAYER          LAYER_NORM
+#   internlm     InternLMDecoderLayer   InternLMRMSNorm
+#   llama 1&2     LlamaDecoderLayer      LlamaRMSNorm
+#    qwen             QWenBlock             RMSNorm
+#   baichuan        DecoderLayer            RMSNorm
+python3 -m lmdeploy.lite.apis.calibrate \
+  --model $HF_MODEL \
+  --layer_type $DECODER_LAYER \  # Decoder Layer 对应的类名
```
Comment (lvhan028) [inline README_zh-CN.md:164]: Can we use model_name to determine what layer_type is?

Comment (pppppM) [reply README_zh-CN.md:164]: Good idea, I'll fix it.

File: lmdeploy/lite/apis/kv_qparams.py [hunk 2]
```diff
@@ -101,18 +101,17 @@ def main(model: str,
...
         'Currently, only support 8-bit quantization for the kv cache.'
     assert calib_dataset in ['c4', 'ptb', 'wikitext2', 'pileval'], \
         'Currently, only support `c4`, `ptb`, `wikitext2`, or `pileval`.'
 
-    tokenizer = AutoTokenizer.from_pretrained(model,
-                                              use_fast=False,
-                                              trust_remote_code=True)
-    model = AutoModelForCausalLM.from_pretrained(model, trust_remote_code=True)
+    tokenizer = AutoTokenizer.from_pretrained(model, use_fast=False)
+    model = AutoModelForCausalLM.from_pretrained(model)
```
Comment (LZHgrla) [inline lmdeploy/lite/apis/kv_qparams.py:114]: Why remove the `trust_remote_code=True`?

File: lmdeploy/lite/apis/kv_qparams.py [hunk 3]
```diff
@@ -124,40 +123,16 @@ def main(model: str,
...
-                          '`nn.Linear`. You can add more robust modules to '
-                          'the `OFFLOAD_MOD_MAP` for faster speed.')
-            offload_mod = OFFLOAD_MOD_MAP[type(model)]
-        with memory_efficient_inference(model, offload_mod):
-            for data in tqdm(calib_loader, desc='Calibrating: '):
-                if isinstance(data, torch.Tensor):
-                    output = model(data.to('cuda'))
-                else:
-                    output = model(data[0].to('cuda'))
-                kv_cache = output.past_key_values
```
Comment (LZHgrla) [inline lmdeploy/lite/apis/kv_qparams.py:146]: use the `KV_CACHE_SIGNATURE` of `lmdeploy/lite/defaults.py`?

File: lmdeploy/lite/quantization/calibration.py [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+      - layer_type: Layer type to be targeted for calibration
+      - norm_type: Normalization type used for calibration
+      - smooth: Flag to indicate if smoothing to be applied or not
+      - w_qconfig: Config file path for weight quantization
+      - work_dir: Directory path to save the calibration and quantization
+                        results
+      - device: Device on which model is to be calibrated ('cpu' or 'cuda')
+
+    Note: More detailed information should be added here to explain what this
+            class does exactly and how it works.
```
Comment (LZHgrla) [inline lmdeploy/lite/quantization/calibration.py:34]: wrong docstring

File: lmdeploy/lite/quantization/weight/quantizer.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+              qparams: Optional[QParams] = None,
+              real: bool = False) -> torch.Tensor:
+        """Perform fake quantization on the given weight tensor.
+
+        Args:
+            weight (torch.Tensor): The weight tensor with shape
+                (out_features, in_features).
+            qparams (Optional[QParams]): A namedtuple containing 'scales'
+                and 'zero_points'.
+            inplace (bool): If True, perform the operation in-place.
```
Comment (LZHgrla) [inline lmdeploy/lite/quantization/weight/quantizer.py:116]:
```suggestion
            real (bool): If True, return the tensor with quantized type.
```

File: lmdeploy/lite/utils/calib_dataloader.py [hunk 6]
```diff
@@ -276,7 +282,12 @@ def get_pileval(tokenizer, nsamples, seed, seqlen=512):
     ], None
 
 
-def get_calib_loaders(name, tokenizer, nsamples=128, seed=0, seqlen=2048):
+def get_calib_loaders(name,
+                      tokenizer,
+                      nsamples=128,
+                      seed=0,
+                      seqlen=2048,
+                      model=''):
```
Comment (LZHgrla) [inline lmdeploy/lite/utils/calib_dataloader.py:290]: model is not used

**Comment by lzhangzz** [comment]
Layout of `.scales` is different from AWQ repo

```
AWQ
model.layers.0.self_attn.q_proj.qzeros             torch.Size([32, 512])          torch.int32
model.layers.0.self_attn.q_proj.scales             torch.Size([32, 4096])         torch.float16

This PR
model.layers.0.self_attn.q_proj.qzeros             torch.Size([32, 512])          torch.int32
model.layers.0.self_attn.q_proj.scales             torch.Size([4096, 32, 1])      torch.float16
```

**Comment by pppppM** [comment]
> Layout of `.scales` is different from AWQ repo
> 
> ```
> AWQ
> model.layers.0.self_attn.q_proj.qzeros             torch.Size([32, 512])          torch.int32
> model.layers.0.self_attn.q_proj.scales             torch.Size([32, 4096])         torch.float16
> 
> This PR
> model.layers.0.self_attn.q_proj.qzeros             torch.Size([32, 512])          torch.int32
> model.layers.0.self_attn.q_proj.scales             torch.Size([4096, 32, 1])      torch.float16
> ```
Some details are not perfectly aligned with awq, and the latest commit has fixed this and now the values and shapes are consistent with the llm-awq results.  @lzhangzz 
img_v2_dc85a3fd-84f3-45c2-a8c2-9637f8e9fc6g

File: lmdeploy/lite/apis/kv_qparams.py (hunk 2 above)
Comment (pppppM) [reply lmdeploy/lite/apis/kv_qparams.py:114]: My bad, there seem to be some issues when rebasing.

File: lmdeploy/lite/utils/calib_dataloader.py (hunk 6 above)
Comment (pppppM) [reply lmdeploy/lite/utils/calib_dataloader.py:290]: My bad, there seem to be some issues when rebasing.

File: lmdeploy/lite/apis/kv_qparams.py (hunk 3 above)
Comment (pppppM) [reply lmdeploy/lite/apis/kv_qparams.py:146]: This is just a temporary solution, and this script will also be deprecated. The quantization parameters of kv cache can be obtained through the `calibrate.py`.

File: .gitignore [hunk 7]
```diff
@@ -1,9 +1,45 @@
...
+eggs/
+.eggs/
+lib/
+lib64/
+parts/
+sdist/
+var/
+wheels/
 *.egg-info/
-workspace/
```
Comment (lvhan028) [inline .gitignore:6]: Why remove workspace?

File: .gitignore [hunk 8]
```diff
@@ -12,4 +48,13 @@ dist/
 examples/cpp/llama/*.csv
 *.npy
 *.weight
-*.pyc
```
Comment (lvhan028) [inline .gitignore:15]: why remove *.pyc

File: .gitignore [hunk 9]
```diff
@@ -1,9 +1,45 @@
...
 __pycache__/
+*.py[cod]
+*$py.class
+
+# C extensions
+*.so
+
+# Distribution / packaging
+.Python
+build/
```
Comment (lvhan028) [inline .gitignore:11]: `*build*/` has been ignored.

File: README_zh-CN.md [hunk 10]
```diff
@@ -151,6 +151,25 @@ deepspeed --module --num_gpus 2 lmdeploy.pytorch.chat \
 
 ## 量化部署
 
+### Weight Only 量化
```
Comment (lvhan028) [inline README_zh-CN.md:154]: May also update README.md

File: lmdeploy/lite/apis/quantize.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    assert key_stats_url.exists()
+    assert value_stats_url.exists()
+
+    key_stats = torch.load(key_stats_url, map_location='cpu')
+    value_stats = torch.load(value_stats_url, map_location='cpu')
+
+    symmetry = kv_sym
+    bits = kv_bits
+
+    # TODO export turbomind weights
```
Comment (lvhan028) [inline lmdeploy/lite/apis/quantize.py:95]: The first 'TODO' can be removed, can't it?

**Comment by lvhan028** [comment]
May update 'News' part in both README

File: .gitignore [hunk 12]
```diff
@@ -1,9 +1,45 @@
-.vscode/
```
Comment (lvhan028) [inline .gitignore:1]: Please keep .vscode and .idea, which are generated by vscode and clion respectively

File: lmdeploy/lite/apis/quantize.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+def main(qparams_dir: str,
+         turbomind_dir: Optional[str],
+         kv_bits: int = 8,
+         kv_sym: bool = False,
+         model_parallel: int = 1):
+
+    qparams_dir = Path(qparams_dir)
+
+    key_stats_url = qparams_dir / 'kv_stats.pth'
```
Comment (humu789) [inline lmdeploy/lite/apis/quantize.py:84]:
```suggestion
    key_stats_url = qparams_dir / 'key_stats.pth'
```

**Review by humu789 — CHANGES_REQUESTED** [review]

File: lmdeploy/lite/apis/calibrate.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+}
+NORM_TYPE_MAP = {
+    'InternLMForCausalLM': 'InternLMRMSNorm',
+    'QWenLMHeadModel': 'RMSNorm',
+    'BaiChuanForCausalLM': 'RMSNorm',
+    'LlamaForCausalLM': 'LlamaRMSNorm',
+}
+
+
+def main(model: str,
```
Comment (lvhan028) [inline lmdeploy/lite/apis/calibrate.py:26]: Is it appropriate to name an API as 'main'?

File: lmdeploy/lite/apis/quantize.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+from pathlib import Path
+from typing import Optional
+
+import fire
+import numpy as np
+import torch
+
+
+def export_turbomind_kv(key_stats, value_stats, bits, symmetry, mp, out_dir):
+
```
Comment (lvhan028) [inline lmdeploy/lite/apis/quantize.py:11]: May update docstring

File: lmdeploy/lite/quantization/activation/observer.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+import torch
+
+from lmdeploy.lite.utils.global_avail import GlobalAvailMixin
+
+
+class KVCacheObserver(GlobalAvailMixin):
+
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/activation/observer.py:9]: May update docstring

File: lmdeploy/lite/quantization/activation/observer.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        cur_min = x.flatten(0, 1).min(0)[0].cpu()
+        cur_absmax = x.flatten(0, 1).abs().max(0)[0].cpu()
+
+        self.max_val = torch.maximum(self.max_val, cur_max)
+        self.min_val = torch.minimum(self.min_val, cur_min)
+        self.absmax_val = torch.maximum(self.absmax_val, cur_absmax)
+
+
+class ActivationObserver(GlobalAvailMixin):
+
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/activation/observer.py:42]: May update docstring

File: lmdeploy/lite/quantization/weight/observer.py [hunk 18]
```diff
@@ -0,0 +1,57 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+import torch
+
+from lmdeploy.lite.utils.global_avail import GlobalAvailMixin
+
+
+class WeightObserver(GlobalAvailMixin):
+
```
Comment (lvhan028) [inline lmdeploy/lite/quantization/weight/observer.py:9]: May update docstring

**Comment by lvhan028** [comment]
The current output is:
```
work_dir/
├── config.json
├── generation_config.json
├── inputs_stats.pth
├── key_stats.pth
├── outputs_stats.pth
├── pytorch_model.bin
└── value_stats.pth
```
I suggest exporting tokenizer-related files from the input huggingface transformers mode to the output directory so that turbomind can deploy the whole package directly. @lzhangzz any ideas?
```
work_dir/
├── config.json
├── generation_config.json
├── inputs_stats.pth
├── key_stats.pth
├── outputs_stats.pth
├── pytorch_model.bin
└── value_stats.pth
├── pytorch_model.bin.index.json
├── special_tokens_map.json
├── tokenizer_config.json
├── tokenizer.json
└── tokenizer.model
```

**Comment by lvhan028** [comment]
May add acknowledgment to llm-awq and related others

**Comment by pppppM** [comment]
I have fixed all comments. Pls review again @lvhan028

**Review by lvhan028 — APPROVED** [review]

**Review by humu789 — APPROVED** [review]
